docs/development/coding-principles.md

```diff
@@ -0,0 +1,140 @@
+    extent.
+
+## Testing
+
+1. All handlers and services should have dedicated unit tests with mocks used for dependencies.
+2. Prefer unit tests over all other test types: unit > integration/component > e2e
+3. When appropriate, prefer composing services via the constructor so that dependencies can be mocked as opposed to
+   instantiating dependencies directly.
+4. Use classes with static test functions to provide common helper methods, e.g. to instantiate an object.
+5. Use `[METHOD]_when[CONDITION]_should[EXPECTATION]` as naming template for test methods,
+   e.g. `verifyInput_whenNull_shouldThrowNpe()` as opposed to `testInputNull()`
+
+## Data Objects
+
+1. Use the `Builder` pattern when:
+    - there are any number of optional constructor args
+    - there are more than 3 constructor args
+    - inheriting from an object that fulfills any of the above. In this case use derived builders as well.
+
+2. Although serializability is not the reason we use the builder pattern, it is a strong indication that a builder
+   should be used.
+2. Builders should be named just `Builder` and be static nested classes.
+3. Create a `public static Builder newInstance(){...}` method to instantiate the builder
+4. Builders have non-public constructors
+5. Use single-field builders: a `Builder` instantiates the object it builds in its constructor, and sets the properties
+   in its builder methods. The `build()` method then only performs verification (optional) and returns the instance.
+6. Use `private` constructors for the objects that the builder builds.
+7. If there is a builder for an object, use it to deserialize an object, i.e. put Jackson annotations such
+   as `JsonCreator` and `@JsonBuilder` on builders.
+8. Note that the motivation behind use of builders is not for immutability (although that may be good in certain
+   circumstances). Rather, it is to make code less error-prone and
+   simpler given the lack of named arguments and optional parameters in Java.
+
+## Secrets
+
+1. Only store secrets in the `Vault` and do not hold them in objects that may be persisted to other stores.
+2. Do not log secrets or sensitive information.
+
+## Extensions and Libraries
+
+1. Extension modules contribute a feature to the runtime such as a service.
+2. SPI modules define extensibility points in the runtime. There is a core SPI module that defines extensibility for
+   essential runtime features. There are other SPI modules that
+   define extensibility points for optional features such as IDS.
+3. Libraries are utility modules that provide classes which may be used by other modules. They do not directly
+   contribute features to the runtime.
+4. An SPI module may only reference other SPI modules and library modules.
+5. An Extension module may only reference other SPI modules and library modules.
+6. A library module may only reference other library modules.
```

Comment:
I am currently considering whether tractusx-edc will provide SPIs in the near future or not.

Any "general" SPI changes (general = not specific to an automotive dataspace connector) would be better contributed to the eclipse-edc than tractusx-edc for better interoperability.

In addition, I don't think there will be any automotive specific functionality that would require SPIs to be introduced.

What are your thoughts on this?